Civi/Api4/Event/Subscriber/CustomValueAccessSubscriber.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+namespace Civi\Api4\Event\Subscriber;
+
+use Civi\API\Events;
+use Civi\Api4\Event\AuthorizeRecordEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Enforce constraints on Custom entities
+ *
+ * Currently:
+ * - deny create if max_multiple reached for the given entity_id
+ *
+ * TODO:
+ * - deny delete if min_multiple under threat
+ * - deny replace / save if it will threaten min/max
+ * - ?
+ *
+ * @service internal
+ */
+class CustomValueAccessSubscriber extends \Civi\Core\Service\AutoService implements EventSubscriberInterface {
+
+  /**
+   * @return array
+   */
+  public static function getSubscribedEvents() {
+    return [
+      'civi.api4.authorizeRecord' => ['onApiAuthorizeRecord', Events::W_LATE],
+    ];
+  }
+
+  /**
+   * @param \Civi\Api4\Event\AuthorizeRecordEvent $event
+   *   API record authorize event
+   */
+  public function onApiAuthorizeRecord(AuthorizeRecordEvent $event) {
+    $apiRequest = $event->getApiRequest();
+
+    if (!in_array(
+        'Civi\Api4\Generic\Traits\CustomValueActionTrait',
+        class_uses($apiRequest)
+        )) {
+      return;
+    }
+
+    $action = $apiRequest->getActionName();
+
+    // TODO: what api actions should we apply to?
+    // min/max should apply to create + delete
+    // also save and replace but trickier
+    // also update if we ever allow changing entity_id
+
+    // for now only apply to create
+    if ($action !== 'create') {
+      return;
+    }
+
+    $parentId = $event->getRecord()['entity_id'] ?? NULL;
+    if (!$parentId) {
+      // we dont know the parent so cant check existing records
+      return;
+    }
+
+    $group = \Civi\Api4\CustomGroup::get(FALSE)
+      ->addSelect('name', 'min_multiple', 'max_multiple')
+      ->addWhere('name', '=', $apiRequest->getCustomGroup())
+      ->addWhere('is_multiple', '=', TRUE)
+      ->execute()->first();
```

Comment:
Custom groups are cached in memory, so we can avoid this database hit by using `CRM_Core_BAO_CustomGroup::getGroup`

Reply:
Ah interesting, I was wondering that.

Reply:
Could CustomGroup::get be taught to use the cache?

Reply:
Otherwise this feels like an issue with the whole BasicBatchAction structure :grimacing:

Reply:
> Could CustomGroup::get be taught to use the cache?

The short answer is: not really. Although in most use-cases swapping the `CustomGroup` api for a `BasicGetAction` that used the cache backend would be not noticeable, all of the SQL functionality would disappear, including joins, group-by, and field transformations, which would break more advanced uses of the api.

> Otherwise this feels like an issue with the whole BasicBatchAction structure 😬

Sorry I must have missed something. Other than swapping out this api call for the cached function, what would be the issue?

Reply:
So my worry is that the original implementation for the `afblockCustom` uses `CRM_Core_BAO_CustomGroup::getAll` 

But I've replaced with two basic batch actions on `CustomGroup::getAfforms` and `CustomGroup::getSearchKit` - which are going to be hitting the database to fetch the records to run through.

It seemed a neat way to structure the code to me, and it's helpful for testing that you can call the Api action directly... but performance-wise maybe not so great?

Reply:
@ufundo I think structuring them as api actions is fine, but what's the problem with those actions internally calling `CRM_Core_BAO_CustomGroup::getAll()` instead of `Civi\Api4\CustomGroup::get()`?